src/promptflow-devkit/promptflow/_sdk/_orm/trace.py

```diff
@@ -234,3 +246,253 @@ def _get_children(line_run_id: str) -> typing.List["LineRun"]:
+    "prompt": "cumulative_token_count",
+    "completion": "cumulative_token_count",
+}
+
+
+@dataclass
+class SearchTransStackItem:
+    orm_op: typing.Union[sqlalchemy.and_, sqlalchemy.or_]
+    expected_length: int
+    conditions: typing.List[sqlalchemy.text]
+
+    def append_condition(self, condition: sqlalchemy.text) -> None:
+        self.conditions.append(condition)
+
+    @property
+    def is_full(self) -> bool:
+        return len(self.conditions) == self.expected_length
+
+    @property
+    def orm_condition(self):
+        return self.orm_op(*self.conditions)
+
+
+class SearchTranslator(ast.NodeVisitor):
+    """Translate line run search to SQLite query."""
+
+    LINE_RUN_SEARCHABLE_FIELDS = [
+        "name",
+        "kind",
+        "status",
+        "start_time",
+        # token count
+        "cumulative_token_count.total",
+        "cumulative_token_count.prompt",
+        "cumulative_token_count.completion",
+        # kind of syntax sugar: hide cumulative_token_count
+        "total",
+        "prompt",
+        "completion",
+        # filter
+        "collection",
+        "run",
+        "session_id",
+    ]
+    LINE_RUN_JSON_FIELDS = {
+        "total": "cumulative_token_count",
+        "prompt": "cumulative_token_count",
+        "completion": "cumulative_token_count",
+    }
+
+    def __init__(self, model):
+        self._model = model
+        # for query build during AST traversal
+        self._stack: typing.List[SearchTransStackItem] = list()
+        self._orm_condition_from_ast = None
+        self._searched_json_fields = set()
+
+    def _build_query(self, session: Session) -> Query:
+        query = session.query(self._model)
+        # if no JSON field is queried, directly build the query
+        if len(self._searched_json_fields) == 0:
+            return query.filter(self._orm_condition_from_ast)
+        # otherwise, manually append "IS NOT NULL"
+        orm_conditions = [self._orm_condition_from_ast]
+        for field in self._searched_json_fields:
+            orm_conditions.append(sqlalchemy.text(f"{field} IS NOT NULL"))
+        return query.filter(sqlalchemy.and_(*orm_conditions))
+
+    def translate(self, session: Session, expression: str) -> Query:
+        # parse expression to AST
+        try:
+            tree = ast.parse(expression, mode="eval")
+            # expression like "name" is not valid
+            assert isinstance(tree.body, (ast.Compare, ast.BoolOp))
+        except (SyntaxError, AssertionError):
+            error_message = "Invalid search expression, currently support Python syntax for search."
+            raise WrongTraceSearchExpressionError(error_message)
+        # traverse the AST and validate the fields are searchable
+        # leveraging `ast.NodeVisitor.visit`
+        self.visit(tree.body)
+        return self._build_query(session=session)
+
+    # override visit BoolOp and Compare methods
+    def visit_BoolOp(self, node: ast.BoolOp) -> None:
```

Comment:
why name not pythonic here?

Reply:
`visit_XXX`, e.g., `visit_BoolOp` is python built-in AST module's function, i have no idea why they are not pythonic, as a python built-in function lol...